nni/common/serializer.py

```diff
@@ -0,0 +1,369 @@
+    def _get_nni_attr(self, name):
+        return self.__dict__['_nni_' + name]
+
+    def __repr__(self):
+        if self._get_nni_attr('self_contained'):
+            return repr(self)
+        if '_nni_cache' in self.__dict__:
+            return repr(self._get_nni_attr('cache'))
+        return 'SerializableObject(' + \
+            ', '.join(['type=' + self._get_nni_attr('symbol').__name__] +
+                      [repr(d) for d in self._get_nni_attr('args')] +
+                      [k + '=' + repr(v) for k, v in self._get_nni_attr('kwargs').items()]) + \
+            ')'
+
+    @staticmethod
+    def _recursive_init(d):
+        # auto-call get() to prevent type-converting in downstreaming functions
+        if isinstance(d, dict):
+            return {k: v.get() if isinstance(v, SerializableObject) else v for k, v in d.items()}
+        else:
+            return [v.get() if isinstance(v, SerializableObject) else v for v in d]
+
+
+def trace(cls_or_func: T = None, /, *, kw_only: bool = True) -> Union[T, SerializableObject]:
+    """
+    Annotate a function or a class if you want to preserve where it comes from.
+    This is usually used in the following scenarios:
+
+    1) Care more about execution configuration rather than results, which is usually the case in AutoML.
+    2) Repeat execution is not an issue (e.g., reproducible, execution is fast without side effects).
+
+    When a class/function is annotated, all the instances/calls will return a object as it normally will.
+    Although the object might act like a normal object, it's actually a different object with NNI-specific properties.
+    To get the original object, you should use ``obj.get()`` to retrieve. The retrieved object can be used
+    like the original one, but there are still subtle differences in implementation.
+
+    Note that when using the result from a trace in another trace-able function/class, ``.get()`` is automatically
+    called, so that you don't have to worry about type-converting.
+
+    Also it records extra information about where this object comes from. That's why it's called "trace".
+    When call ``nni.dump``, that information will be used, by default.
+
+    If ``kw_only`` is true, try to convert all parameters into kwargs type. This is done by inspect the argument
+    list and types. This can be useful to extract semantics, but can be tricky in some corner cases.
+
+    Example:
+
+    .. code-block:: python
+
+        @nni.trace
+        def foo(bar):
+            pass
+    """
+
+    def wrap(cls_or_func):
+        if isinstance(cls_or_func, type):
+            return _trace_cls(cls_or_func, kw_only)
+        else:
+            return _trace_func(cls_or_func, kw_only)
+
+    # if we're being called as @trace()
+    if cls_or_func is None:
+        return wrap
+
+    # if we are called without parentheses
+    return wrap(cls_or_func)
+
+
+def dump(obj: Any, fp: Optional[Any] = None, use_trace: bool = True, pickle_size_limit: int = 4096,
```

Comment:
what are `dump` and `load` used for?

Reply:
Same semantics as dump and load in json-tricks.